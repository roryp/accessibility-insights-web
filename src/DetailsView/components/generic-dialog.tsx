// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { DefaultButton, PrimaryButton } from 'office-ui-fabric-react';
import { Dialog, DialogFooter, DialogType } from 'office-ui-fabric-react';
import * as React from 'react';

import { NamedFC } from '../../common/react/named-fc';

export type GenericDialogProps = {
    onPrimaryButtonClick: (event: React.MouseEvent<any>) => void;
    onCancelButtonClick: (event: React.MouseEvent<any>) => void;
    messageText: string;
    title: string;
    primaryButtonText: string;
};

export const GenericDialog = NamedFC<GenericDialogProps>('GenericDialog', props => {
    const {
        onCancelButtonClick,
        onPrimaryButtonClick,
        messageText,
        title,
        primaryButtonText,
    } = props;

    return (
        <Dialog
            hidden={false}
            onDismiss={onCancelButtonClick}
            dialogContentProps={{
                type: DialogType.normal,
                title: title,
                showCloseButton: false,
            }}
            modalProps={{
                isBlocking: false,
                containerClassName: 'insights-dialog-main-override',
            }}
        >
            <div className={'start-over-dialog-body'}>{messageText}</div>
            <DialogFooter>
                <PrimaryButton onClick={onPrimaryButtonClick} text={primaryButtonText} />
                <DefaultButton onClick={onCancelButtonClick} text={'Cancel'} autoFocus={true} />
            </DialogFooter>
        </Dialog>
    );
});
