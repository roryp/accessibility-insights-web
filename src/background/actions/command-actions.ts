// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Action } from 'common/flux/action';

export interface GetCommandsPayload {
    commands: chrome.commands.Command[];
    tabId: number;
}

export class CommandActions {
    public readonly getCommands = new Action<GetCommandsPayload>();
}
