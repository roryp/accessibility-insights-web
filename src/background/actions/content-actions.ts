// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Action } from 'common/flux/action';
import { BaseActionPayload } from './action-payloads';

export interface ContentPayload extends BaseActionPayload {
    contentPath: string;
    contentTitle: string;
}

export class ContentActions {
    public readonly openContentPanel = new Action<ContentPayload>();
    public readonly closeContentPanel = new Action<void>();
}
