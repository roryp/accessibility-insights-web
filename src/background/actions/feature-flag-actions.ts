// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Action } from 'common/flux/action';

import { BaseActionPayload } from './action-payloads';

export interface FeatureFlagPayload extends BaseActionPayload {
    feature: string;
    enabled: boolean;
}

export class FeatureFlagActions {
    public readonly getCurrentState = new Action<void>();
    public readonly setFeatureFlag = new Action<FeatureFlagPayload>();
    public readonly resetFeatureFlags = new Action<void>();
}
