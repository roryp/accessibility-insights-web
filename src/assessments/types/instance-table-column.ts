// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { AssessmentInstanceRowData } from 'DetailsView/components/assessment-instance-table';

export interface InstanceTableColumn {
    key: string;
    name: string;
    onRender: (item: AssessmentInstanceRowData) => JSX.Element;
}
