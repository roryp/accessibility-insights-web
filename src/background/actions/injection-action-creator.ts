// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Messages } from 'common/messages';
import { InjectionActions } from '../actions/injection-actions';
import { Interpreter } from '../interpreter';

export class InjectionActionCreator {
    constructor(
        private readonly interpreter: Interpreter,
        private readonly injectionActions: InjectionActions,
    ) {}

    public registerCallbacks(): void {
        this.interpreter.registerTypeToPayloadCallback(
            Messages.Visualizations.State.InjectionStarted,
            this.injectionStarted,
        );
        this.interpreter.registerTypeToPayloadCallback(
            Messages.Visualizations.State.InjectionCompleted,
            this.injectionCompleted,
        );
    }

    private injectionStarted = (): void => this.injectionActions.injectionStarted.invoke(null);

    private injectionCompleted = (): void => this.injectionActions.injectionCompleted.invoke(null);
}
