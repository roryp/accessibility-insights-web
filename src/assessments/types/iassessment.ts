// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { InitialDataCreator } from 'background/create-initial-assessment-test-data';
import { RequirementOrdering } from 'common/assessment/requirement';
import { AssessmentVisualizationConfiguration } from 'common/configs/assessment-visualization-configuration';
import { AnyExtension } from 'common/extensibility/extension-point';
import { VisualizationType } from 'common/types/visualization-type';
import { ContentPageComponent } from 'views/content/content-page';
import { Requirement } from './requirement';

interface BaseAssessment {
    key: string;
    visualizationType: VisualizationType;
    title: string;
    gettingStarted: JSX.Element;
    guidance?: ContentPageComponent;
    requirements: Requirement[];
    featureFlag?: { required?: string[] };
    requirementOrder?: RequirementOrdering;
    extensions?: AnyExtension[];
    initialDataCreator?: InitialDataCreator;
}

export interface ManualAssessment extends BaseAssessment {}

export interface AssistedAssessment extends BaseAssessment {
    storeDataKey: string;
    visualizationConfiguration?: Partial<AssessmentVisualizationConfiguration>;
}

export interface Assessment extends BaseAssessment {
    getVisualizationConfiguration: () => AssessmentVisualizationConfiguration;
    requirementOrder: RequirementOrdering;
}
