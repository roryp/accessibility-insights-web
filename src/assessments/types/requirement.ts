// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { UniquelyIdentifiableInstances } from 'background/instance-identifier-generator';
import { ManualTestStatus } from 'common/types/manual-test-status';
import {
    AssessmentNavState,
    GeneratedAssessmentInstance,
    InstanceIdToInstanceDataMap,
} from 'common/types/store-data/assessment-result-data';
import { FeatureFlagStoreData } from 'common/types/store-data/feature-flag-store-data';
import { DetailsViewActionMessageCreator } from 'DetailsView/actions/details-view-action-message-creator';
import {
    AssessmentInstanceRowData,
    AssessmentInstanceTable,
} from 'DetailsView/components/assessment-instance-table';
import { RequirementLink } from 'DetailsView/components/requirement-link';
import { Analyzer } from 'injected/analyzers/analyzer';
import { AnalyzerProvider } from 'injected/analyzers/analyzer-provider';
import { DecoratedAxeNodeResult } from 'injected/scanner-utils';
import {
    PropertyBags,
    VisualizationInstanceProcessorCallback,
} from 'injected/visualization-instance-processor';
import { Drawer } from 'injected/visualization/drawer';
import { DrawerProvider } from 'injected/visualization/drawer-provider';
import { IColumn } from 'office-ui-fabric-react';
import { DictionaryStringTo } from 'types/common-types';
import { ContentPageComponent, HyperlinkDefinition } from 'views/content/content-page';
import { IGetMessageGenerator } from '../assessment-default-message-generator';
import { InstanceTableColumn } from './instance-table-column';
import { ReportInstanceFields } from './report-instance-field';

export interface Requirement {
    key: string;
    description: JSX.Element;
    order?: number;
    name: string;
    renderStaticContent?: () => JSX.Element;
    howToTest: JSX.Element;
    addFailureInstruction?: string;
    infoAndExamples?: ContentPageComponent;
    isManual: boolean;
    // This is for semi-manual cases where we can't present a list of instances like an assisted
    // test would, but can infer a PASS or FAIL state. If not specified, acts like () => UNKNOWN.
    getInitialManualTestStatus?: (instances: InstanceIdToInstanceDataMap) => ManualTestStatus;
    guidanceLinks: HyperlinkDefinition[];
    columnsConfig?: InstanceTableColumn[];
    getAnalyzer?: (provider: AnalyzerProvider) => Analyzer;
    getVisualHelperToggle?: (props: VisualHelperToggleConfig) => JSX.Element;
    // Any results this returns false for will be omitted from visual helper displays, but still
    // present for the purposes of instance lists or getInitialManualTestStatus. By default, all
    // results support visualization.
    isVisualizationSupportedForResult?: (result: DecoratedAxeNodeResult) => boolean;
    visualizationInstanceProcessor?: VisualizationInstanceProcessorCallback<
        PropertyBags,
        PropertyBags
    >;
    getDrawer?: (provider: DrawerProvider, featureFlagStoreData?: FeatureFlagStoreData) => Drawer;
    getNotificationMessage?: (selectorMap: DictionaryStringTo<any>) => string;
    doNotScanByDefault?: boolean;
    switchToTargetTabOnScan?: boolean;
    generateInstanceIdentifier?: (instance: UniquelyIdentifiableInstances) => string;
    reportInstanceFields?: ReportInstanceFields;
    renderReportDescription?: () => JSX.Element;
    getInstanceStatus?: (result: DecoratedAxeNodeResult) => ManualTestStatus;
    getInstanceStatusColumns?: () => Readonly<IColumn>[];
    getDefaultMessage?: IGetMessageGenerator;
    renderInstanceTableHeader?: (
        table: AssessmentInstanceTable,
        items: AssessmentInstanceRowData[],
    ) => JSX.Element;
    renderRequirementDescription?: (requirementLink: RequirementLink) => JSX.Element;
}

export type VisualHelperToggleConfigDeps = {
    detailsViewActionMessageCreator: DetailsViewActionMessageCreator;
};

export interface VisualHelperToggleConfig {
    deps: VisualHelperToggleConfigDeps;
    assessmentNavState: AssessmentNavState;
    instancesMap: DictionaryStringTo<GeneratedAssessmentInstance>;
    isStepEnabled: boolean;
    isStepScanned: boolean;
}
