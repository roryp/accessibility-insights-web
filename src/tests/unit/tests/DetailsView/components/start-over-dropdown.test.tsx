// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { shallow } from 'enzyme';
import { ContextualMenu } from 'office-ui-fabric-react';
import * as React from 'react';
import { IMock, Mock, Times } from 'typemoq';

import { InsightsCommandButton } from 'common/components/controls/insights-command-button';
import { VisualizationType } from '../../../../../common/types/visualization-type';
import { DetailsViewActionMessageCreator } from '../../../../../DetailsView/actions/details-view-action-message-creator';
import { DetailsRightPanelConfiguration } from '../../../../../DetailsView/components/details-view-right-panel';
import { GenericDialog } from '../../../../../DetailsView/components/generic-dialog';
import {
    StartOverDropdown,
    StartOverProps,
} from '../../../../../DetailsView/components/start-over-dropdown';

describe('StartOverDropdownTest', () => {
    let defaultProps: StartOverProps;
    let detailsViewActionMessageCreatorMock: IMock<DetailsViewActionMessageCreator>;

    const event = {
        currentTarget: 'test target',
    } as React.MouseEvent<any>;

    beforeEach(() => {
        detailsViewActionMessageCreatorMock = Mock.ofType<DetailsViewActionMessageCreator>();
        defaultProps = {
            deps: {
                detailsViewActionMessageCreator: detailsViewActionMessageCreatorMock.object,
            },
            testName: 'test name',
            test: -1 as VisualizationType,
            requirementKey: 'test key',
            rightPanelConfiguration: {
                GetStartOverContextualMenuItemKeys: () => ['assessment', 'test'],
            } as DetailsRightPanelConfiguration,
            dropdownDirection: 'down',
        };
    });

    it('render', () => {
        const rendered = shallow(<StartOverDropdown {...defaultProps} />);

        expect(rendered.debug()).toMatchSnapshot();
    });

    it('render ContextualMenu', () => {
        const rendered = shallow<StartOverDropdown>(<StartOverDropdown {...defaultProps} />);
        rendered.find(InsightsCommandButton).simulate('click', event);
        expect(rendered.getElement()).toMatchSnapshot();
        expect(rendered.state().target).toBe(event.currentTarget);
    });

    it('render ContextualMenu with only one option', () => {
        defaultProps.rightPanelConfiguration = {
            GetStartOverContextualMenuItemKeys: () => ['assessment'],
        } as DetailsRightPanelConfiguration;
        const rendered = shallow<StartOverDropdown>(<StartOverDropdown {...defaultProps} />);
        rendered.find(InsightsCommandButton).simulate('click', event);
        expect(rendered.getElement()).toMatchSnapshot();
        expect(rendered.state().target).toBe(event.currentTarget);
    });

    it('render GenericDialog for start over a test', () => {
        const rendered = shallow(<StartOverDropdown {...defaultProps} />);
        rendered.find(InsightsCommandButton).simulate('click', event);
        rendered
            .find(ContextualMenu)
            .prop('items')
            .find(elem => elem.key === 'test')
            .onClick();
        expect(rendered.debug()).toMatchSnapshot();
    });

    it('render GenericDialog for start over the whole assessment', () => {
        const rendered = shallow(<StartOverDropdown {...defaultProps} />);
        rendered.find(InsightsCommandButton).simulate('click', event);
        rendered
            .find(ContextualMenu)
            .prop('items')
            .find(elem => elem.key === 'assessment')
            .onClick();
        expect(rendered.debug()).toMatchSnapshot();
    });

    it('render with dropdown on left', () => {
        const props: StartOverProps = {
            ...defaultProps,
            dropdownDirection: 'left',
        };
        const rendered = shallow(<StartOverDropdown {...props} />);
        rendered.find(InsightsCommandButton).simulate('click', event);
        expect(rendered.getElement()).toMatchSnapshot();
    });

    it('should dismiss the start test over dialog', () => {
        detailsViewActionMessageCreatorMock
            .setup(creator =>
                creator.cancelStartOver(event, defaultProps.test, defaultProps.requirementKey),
            )
            .verifiable(Times.once());

        const rendered = shallow<StartOverDropdown>(<StartOverDropdown {...defaultProps} />);
        rendered.find(InsightsCommandButton).simulate('click', event);
        rendered
            .find(ContextualMenu)
            .prop('items')
            .find(elem => elem.key === 'test')
            .onClick();
        rendered.find(GenericDialog).prop('onCancelButtonClick')(event);

        expect(rendered.state().dialogState).toEqual('none');
        detailsViewActionMessageCreatorMock.verifyAll();
    });

    it('should dismiss the start assessment over dialog', () => {
        detailsViewActionMessageCreatorMock
            .setup(creator => creator.cancelStartOverAllAssessments(event))
            .verifiable(Times.once());

        const rendered = shallow<StartOverDropdown>(<StartOverDropdown {...defaultProps} />);
        rendered.find(InsightsCommandButton).simulate('click', event);
        rendered
            .find(ContextualMenu)
            .prop('items')
            .find(elem => elem.key === 'assessment')
            .onClick();
        rendered.find(GenericDialog).prop('onCancelButtonClick')(event);

        expect(rendered.state().dialogState).toEqual('none');
        detailsViewActionMessageCreatorMock.verifyAll();
    });

    it('should start over a test', () => {
        detailsViewActionMessageCreatorMock
            .setup(creator => creator.startOverTest(event, defaultProps.test))
            .verifiable(Times.once());

        const rendered = shallow<StartOverDropdown>(<StartOverDropdown {...defaultProps} />);
        rendered.find(InsightsCommandButton).simulate('click', event);
        rendered
            .find(ContextualMenu)
            .prop('items')
            .find(elem => elem.key === 'test')
            .onClick();
        rendered.find(GenericDialog).prop('onPrimaryButtonClick')(event);

        expect(rendered.state().dialogState).toEqual('none');
        detailsViewActionMessageCreatorMock.verifyAll();
    });

    it('should start over the whole assessment', () => {
        detailsViewActionMessageCreatorMock
            .setup(creator => creator.startOverAllAssessments(event))
            .verifiable(Times.once());

        const rendered = shallow<StartOverDropdown>(<StartOverDropdown {...defaultProps} />);
        rendered.find(InsightsCommandButton).simulate('click', event);
        rendered
            .find(ContextualMenu)
            .prop('items')
            .find(elem => elem.key === 'assessment')
            .onClick();
        rendered.find(GenericDialog).prop('onPrimaryButtonClick')(event);

        expect(rendered.state().dialogState).toEqual('none');
        detailsViewActionMessageCreatorMock.verifyAll();
    });

    it('should dismiss the contextMenu', () => {
        const rendered = shallow<StartOverDropdown>(<StartOverDropdown {...defaultProps} />);
        rendered.find(InsightsCommandButton).simulate('click', event);
        rendered.find(ContextualMenu).prop('onDismiss')();

        expect(rendered.state().isContextMenuVisible).toBe(false);
        expect(rendered.state().target).toBeNull();
    });
});
