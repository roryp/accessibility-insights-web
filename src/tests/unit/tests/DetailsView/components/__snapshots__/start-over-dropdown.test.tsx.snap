// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartOverDropdownTest render 1`] = `
"<div>
  <InsightsCommandButton iconProps={{...}} text=\\"Start over\\" ariaLabel=\\"start over menu\\" onClick={[Function]} menuIconProps={{...}} />
</div>"
`;

exports[`StartOverDropdownTest render ContextualMenu 1`] = `
<div>
  <InsightsCommandButton
    ariaLabel="start over menu"
    iconProps={
      Object {
        "iconName": "Refresh",
      }
    }
    menuIconProps={
      Object {
        "iconName": "ChevronDown",
      }
    }
    onClick={[Function]}
    text="Start over"
  />
  <StyledWithResponsiveMode
    directionalHint={7}
    items={
      Array [
        Object {
          "key": "assessment",
          "name": "Start over Assessment",
          "onClick": [Function],
        },
        Object {
          "key": "test",
          "name": "Start over test name",
          "onClick": [Function],
        },
      ]
    }
    onDismiss={[Function]}
    target="test target"
  />
</div>
`;

exports[`StartOverDropdownTest render ContextualMenu with only one option 1`] = `
<div>
  <InsightsCommandButton
    ariaLabel="start over menu"
    iconProps={
      Object {
        "iconName": "Refresh",
      }
    }
    menuIconProps={
      Object {
        "iconName": "ChevronDown",
      }
    }
    onClick={[Function]}
    text="Start over"
  />
  <StyledWithResponsiveMode
    directionalHint={7}
    items={
      Array [
        Object {
          "key": "assessment",
          "name": "Start over Assessment",
          "onClick": [Function],
        },
      ]
    }
    onDismiss={[Function]}
    target="test target"
  />
</div>
`;

exports[`StartOverDropdownTest render GenericDialog for start over a test 1`] = `
"<div>
  <InsightsCommandButton iconProps={{...}} text=\\"Start over\\" ariaLabel=\\"start over menu\\" onClick={[Function]} menuIconProps={{...}} />
  <StyledWithResponsiveMode onDismiss={[Function: onDismiss]} target=\\"test target\\" items={{...}} directionalHint={7} />
  <GenericDialog title=\\"Start over\\" messageText=\\"Starting over will clear all existing results from the test name test. Are you sure you want to start over?\\" onCancelButtonClick={[Function]} onPrimaryButtonClick={[Function]} primaryButtonText=\\"Start over\\" />
</div>"
`;

exports[`StartOverDropdownTest render GenericDialog for start over the whole assessment 1`] = `
"<div>
  <InsightsCommandButton iconProps={{...}} text=\\"Start over\\" ariaLabel=\\"start over menu\\" onClick={[Function]} menuIconProps={{...}} />
  <StyledWithResponsiveMode onDismiss={[Function: onDismiss]} target=\\"test target\\" items={{...}} directionalHint={7} />
  <GenericDialog title=\\"Start over\\" messageText=\\"Starting over will clear all existing results from the Assessment. This will clear results and progress of all tests and requirements. Are you sure you want to start over?\\" onCancelButtonClick={[Function]} onPrimaryButtonClick={[Function]} primaryButtonText=\\"Start over\\" />
</div>"
`;

exports[`StartOverDropdownTest render with dropdown on left 1`] = `
<div>
  <InsightsCommandButton
    ariaLabel="start over menu"
    iconProps={
      Object {
        "iconName": "Refresh",
      }
    }
    menuIconProps={
      Object {
        "iconName": "ChevronRight",
      }
    }
    onClick={[Function]}
    text="Start over"
  />
  <StyledWithResponsiveMode
    directionalHint={8}
    items={
      Array [
        Object {
          "key": "assessment",
          "name": "Start over Assessment",
          "onClick": [Function],
        },
        Object {
          "key": "test",
          "name": "Start over test name",
          "onClick": [Function],
        },
      ]
    }
    onDismiss={[Function]}
    target="test target"
  />
</div>
`;
